flow/environments/xsede.py

```diff
@@ -215,8 +215,69 @@ def add_args(cls, parser):
         )
 
 
+class ExpanseEnvironment(DefaultSlurmEnvironment):
+    """Environment profile for the Expanse supercomputer.
+
+    https://www.sdsc.edu/support/user_guides/expanse.html
+    """
+
+    hostname_pattern = r".*\.expanse\.sdsc\.edu$"
+    template = "expanse.sh"
+    cores_per_node = 128
+    gpus_per_node = 4
+
+    @classmethod
+    def _get_mpi_prefix(cls, operation):
+        if operation.directives["nranks"] == 0:
+            return ""
+        # We must specify the pmi2 option as the SLURM configuration of Expanse uses the
+        # special pmi2 preprocessing. See `man srun` on an Expanse login node.
+        base_prefix = "srun --mpi=pmi2"
+        rank_option = f"-n {operation.directives['nranks']} "
+        # We have to handle the case with and without OpenMP differently due to the
+        # --cpus-per-task option.
+        if operation.directives["omp_num_threads"] == 0:
+            return base_prefix + rank_option
+        cpus_per_task = f"--cpus-per-task={operation.directives['omp_num_threads']} "
+        return base_prefix + cpus_per_task + rank_option
```

Comment:
Are the MPI and OMP prefixes stacked? If not, this also needs to set the OMP_NUM_THREADS environment variable.

Reply:
Yes, this is done automatically.